### ISSUE TITLE
Default instance settings not being used

### ISSUE THREAD

**Issue by t-schroeder** [Opencast-Moodle/moodle-tool_opencast#45]
We upgraded from an earlier version of the plugin where you could not configure multiple opencast instances. Now we still have the old settings like `apiusername` in addition to the new settings like `apiusername_1`. If we construct an `api` object without passing any parameters, the expected behavior would be for it to use the settings from the opencast instance that is configured as default. Instead it uses the old settings that don't have an ocinstanceid suffix. Those settings are not accessible anymore through the Moodle website. On fresh installs it will probably not even work if you don't explicitly provide an instanceid, since the settings with the old identifier will not exist there.

https://github.com/Opencast-Moodle/moodle-tool_opencast/blob/fe2d5d7107aac4884022d61e83e6c4b441b87152/classes/local/api.php#L165-L177

**Edit:** I just noticed if you pass the instanceid of the default instance, it will also use the old settings, not actually the settings for the instanceid you passed.

## Expected behavior
Constructing an `api` object without providing an instanceid should use the settings from the default instance. If you pass an instanceid, the settings for that instance should be used.

## Actual behavior
It uses the old settings instead if you pass none or the default instanceid. These settings are no longer accessible and don't even exist on new installs.

**Comment by TamaraGunkel** [comment]
Hey,
if you don't have access anymore to the "old" settings without an instance ID, something has gone wrong during your update.
Let me explain the issue in more detail:
- When I implemented the database change, I didn't wanted it to be a breaking change but ensure backwards compatibility. I couldn't remove the old settings without the ID because the plugins relied on them and directly used them. If I would have changed them, the new tool version would not have worked with any older plugin release. 
- During the update, the settings are therefore kept and the old (and only instance) is marked as default. The default instance is connected with the settings without instanceid. If you're using only one instance, you should thus be able to edit them in your Moodle instance and everything should work fine. The update worked for the other universities, so I don't know what the problem is.
- For new instances, the code works also fine. Assuming that new instances use the newest versions of the plugins, the instanceid is always passed to the api. You're right, probably it does not work when the instanceid is not passed. But since this it not the case, it does not really matter.
- There is however a problem when connecting the default instance with the "old" settings. If the default instance is changed, the settings suddenly belong to a wrong instance. This should definitely be fixed in the future (I'm going to open a new issue for that). But since only a few universities use multiple instances and the default instance is usually not changed, it's not such a time critical task.
- I think that we can discard the "old" settings at some point in the future because the new plugin versions do not directly use these settings anymore. And then we can and should tidy up the code. Maybe the upgrade to Moodle 4 would be suitable for doing that. However, as long as we want to keep the backwards compatibility, this is, unfortunately, not possible. 

So basically, this is not fixable because the default instance is directly connected with the old settings. You should check your Moodle configuration. If the instance is marked as default, you should have access to these settings.

**Comment by t-schroeder** [comment]
Thanks for the thorough explanation.

Some background: On our normal Moodle instances we only have one Opencast instance configured each, so there is also no problem for us there. I was just testing around a bit on my development machine with multiple opencast instances. But you are right, the actual problem seems to be that I'm calling the api class constructor with no parameters. We've made some local changes to block_opencast that I forgot to adapt. Once I made sure to pass ocinstanceid to the constructor I no longer ran into this problem. Closing this issue.
